Docs: design for announcing IP from specified interfaces for layer2 (#1359)
metallb/metallb: design/layer2-bind-interfaces.md

## Details

In the previous section, we described the data model in form of Go structures. Now we will use CRs combined with scenarios 
to illustrate the design.

### Announcing VIP from specific interfaces to avoid network connection failure

In some complex physical network environments, spamming ARP broadcasts on all interfaces will cause a loss of connectivity.

#### Announcing all VIPs from specific interfaces

We can configure all VIPs to be announced from a specific network interface on all nodes, by setting interface names 
without specifying IPAddressPools and nodes.
There is an example:
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement2
  namespace: metallb-system
spec:
  interfaces:
    - eth1
```

#### Announcing the different VIPs from different interfaces

VIPs belonging to different subnets could advertise to the corresponding Layer 2 network by specifying correspond interface.
The IPAddressPool named "pool1" belongs to subnet 192.172.1.0/24, and "pool2" belongs to subnet 192.168.1.0/24
```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: pool1
  namespace: metallb-system
spec:
  addresses:

>     - 192.172.1.10-192.172.1.70

Author: 192.172.1.0/24 include the veth1 IP 192.172.1.1. 192.172.1.1 can't be allocated to Service，So use the subset of 192.172.1.0/24 to fill the AddressPool

---
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: pool2
  namespace: metallb-system
spec:
  addresses:
    - 192.168.1.128/26
```
In the following YAML, we can configure the VIP in the subnet 192.172.1.0/24 to publish from veth1 (192.172.1.1) and the VIP in the subnet
192.168.1.0/24 published from veth0(192.168.1.1)
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement3
  namespace: metallb-system
spec:
  ipAddressPools:
    - pool1
  interfaces:
    - veth1
---
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement4
  namespace: metallb-system
spec:
  ipAddressPools:
    - pool2
  interfaces:
    - veth0
```

### Avoid spamming ARP to veth interfaces

> Specifies that the VIP is advertised from all interfaces, except the virtual interfaces that connect to the pod in 
> the form of veth pair. For instance, each node in the cluster has 3 physical interfaces: eth0, eth1, eth2, then the YAML like 
> bellow example.

Reviewer: ```suggestion
Specify that all VIPs are advertised from all physical interfaces, except the virtual interfaces that belong to the pods in the form of veth pair. For example, if each node in the cluster has 3 physical interfaces: eth0, eth1, eth2:
```

```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
name: example-advertisement5
namespace: metallb-system
spec:
  interfaces:
    - eth0
    - eth1
    - eth2
```

### Announcing the VIP from specific interfaces only from a subset of nodes

> - In the YAML below, MetalLB will only announce pool1 IPs from hostA and hostB's interface ens18.

Reviewer: ```suggestion
By mixing `nodeSelectors` and `interfaces`, we can consider announcing pool1 from hostA and hostB, using the selected host's ens18 interface:
```

```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement6
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
  nodeSelector:
    - matchExpression:
      - key: kubernetes.io/hostname
        operator: In
        values: [hostA, hostB]
  interfaces:
    - ens18
```

### Specify different interfaces for different nodes to announce VIP

> In a heterogeneous cluster, the interfaces of each node are not completely consistent. We need to specify different 
> interfaces for different nodes.
> For example, listen on eno1 on "worker" nodes, but enp3s5 and vlan6 on "gateway" nodes.

Reviewer: ```suggestion
In a heterogeneous cluster, the interfaces of each node are not completely consistent. We need to specify different interfaces for different nodes.
For example, when announcing from "worker" nodes use only eno1 interface, but only enp3s5 and vlan6 on "gateway" nodes.
```

# Configuring announcing interfaces

## Summary

The purpose of the design is to support configuring the listening nodes and interfaces for each IPAddressPool for L2 mode.

Previous discussions about this topic can be found in [#277](https://github.com/metallb/metallb/issues/277).

## Motivation

There are two motivations:

The first one is that VIP's ARP broadcasts on all interfaces cause loss of network connectivity in some complex scenarios.

> The picture below shows one of my actual scenarios. Virtual interfaces veth0 and veth1 bridge on one physical interface eth0.

Reviewer 2: I'd put the picture first, and the explanation below. Also, I'd split the text in paragraphs with new lines.

The veth0 is belongs to subnet 192.168.1.0/24, the veth1 is belongs to subnet 192.172.1.0/24. When the Client(192.172.1.2)
requests its LB Service(VIP is 192.172.1.10), it may send packets to veth0(due to the Speaker advertising VIP from all 

> interfaces), but K8s would reply on veth1, which led to asymmetric routing so that Linux drop the packets from the Client.

Reviewer 2: ```suggestion
interfaces), but K8s would reply on veth1, which would lead to asymmetric routing so that Linux drop the packets from the Client.
```

> interfaces), but K8s would reply on veth1, which led to asymmetric routing so that Linux drop the packets from the Client.

Author: The K8s host will drop the requests from the client.

> The VIPs should listen to the suitable interfaces according to the network environment, other than all interfaces.

Reviewer 2: ```suggestion
In order to solve this issue, we should advertise the loadbalancer ip only to a subset of the interfaces according to the network environment, other than all interfaces.
```

[<img src="./images/complex_system.svg" width="800"/>](./images/complex_system.svg)

Secondly, in a system that clearly distinguishes network planes, people hope to publish VIPs on the corresponding network planes 
according to the LB Service function, rather than indiscriminately announcing VIPs on each network plane. For example, in the 
following picture, if one LB Service is only requested by the `management Network`, we hope the Speaker only advertises this VIP 

> on eth0. Of course, this is not necessary.

Reviewer 2: I'd mention "Failover ARP spam broadcasts on all interfaces. While again, this is not a security problem in any way, it can confuse dumb networking gear with broken firmware, causing loss of connectivity." as per https://github.com/metallb/metallb/issues/277#issuecomment-559598076

[<img src="./images/distinguish_network_plane.png" width="800"/>](./images/distinguish_network_plane.png)

### Goals

> - In L2 mode, specify an interfaces list used by MetalLB to announce VIPs, applicable to all nodes and IPAddressPools.

Reviewer 2: ```suggestion
- being able to specify a subset of interfaces used by MetalLB to announce loadbalancer ips, applicable to all nodes and IPAddressPools.
```

Reviewer 2: Same for below rows, I'd just mention in a sentence that this applies to L2 mode only and I'd remove the i.e. part

  (i.e. support config `(Interfaces)`)
- In L2 mode, specify an interfaces list used by MetalLB to announce some specified IPAddressPools' IPs, applicable to all 
  nodes. (i.e. support config `(IPAddressPools, Interfaces)`)
- In L2 mode, specify an interfaces list used by MetalLB to announce some specified IPAddressPools' IPs, applicable to the 
  specified nodes. (i.e. support config `(  IPAddressPools, Nodes, Interfaces)`)

> ### Non-Goals

Reviewer 2: nit: add a space

- Determine the interface used to announce the VIP according to the host route.
- Attempt to find and correct user's configuration errors. For example, if the user-configured interface's name does not exist, 

>   MetalLB will not announce the VIP instead of using another interface to announce the VIP.

Reviewer 2: I'd add: "influence the leader election mechanism for each loadbalancerIP". If one of the nodes selected becomes the leader for that particular IP, and no interfaces are available, MetalLB will not choose another node.
(I think it makes sense to throw at least an event though)

## Data model

Based on the current MetalLB CRD design (`8d906f0abef3d7ed5b0435fa4edfbe5e37b38557`), we'd like to add a new field `Interfaces` to `L2Advertisement`.

The new definition of `L2Advertisement` is:
```go
type L2Advertisement struct {
  Name                     string `yaml:"name"`
  IPAddressPools           []string `yaml:"ipAddressPools"`
  IPAddressPoolSelector    *metav1.LabelSelector `yaml:"ipAddressPoolSelector"`
  NodeSelector             *metav1.LabelSelector `yaml:"nodeSelector"`
  Interfaces               []string `yaml:"interfaces"`
}
```

> - `Interfaces`: It's an interface name list. The VIP in IPAddressPools(specified by `IPAddressPools` or `IPAddressPoolSelector`) only be announced 

Reviewer 2: ```suggestion
- `Interfaces`: It's an interface name list. The LoadBalancerIP of a service belonging to the IPAddressPools selected by this `L2Advertisement` will be announced only by the selected interfaces.

If the interfaces field is empty, the LoadBalancerIP will be advertised via all the interfaces.
```

  by the `Interfaces` of the nodes(specified by `NodeSelector`). If the field `IPAddressPools` and `IPAddressPoolSelector` 

>   is null, the `Interfaces` applicable to all IPAddressPools. If the field `NodeSelector` is null, the `Interfaces` 

Reviewer 2: I'd remove the nodeselector paragraph, as it's part of the existing mechanism and nothing changes.

  applicable to all nodes.

## Details

> In the previous section, we described the data model in form of Go structures. Now we will use CR combined with scenarios 

Reviewer 2: ```suggestion
In the previous section, we described the data model in form of Go structures. Now we will use CRs combined with scenarios 
```

to illustrate the design.

### Compatible with existing logic

- The design does not change the original logic and supports upgrade. If you don't set field `Interface`, nothing has changed.

> An example definition can be:

Author: I has removed the example

```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement1
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
    - pool2
  nodeSelector:
    - matchExpression:
      - key: kubernetes.io/hostname
        operator: In
        values: [hostA, hostB]
```

### Announcing VIP from specific interfaces to avoid network connection failure

> In some complex physical network environments, ARP spam broadcasts on all interfaces cause a loss of connectivity.

Reviewer 2: ```suggestion
In some complex physical network environments, spamming ARP broadcasts on all interfaces will cause a loss of connectivity.
```

> - We can configure all VIPs to be announced from a special network port on all nodes. don't specify IPAddressPools and nodes, 

Reviewer 2: ```suggestion
- We can configure all VIPs to be announced from a specific network interface on all nodes, without specifying IPAddressPools and nodes by setting the interface name.
```

Author: I has changed the order of phrases

only set interface names
There is an example:
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement2
  namespace: metallb-system
spec:
  interfaces:
    - eth1
```

> - VIPs belonging to different subnets could advertise to the corresponding Layer 2 network by specifying correspond interface.

Reviewer 2: This should be a new subtitle?

The IPAddressPool named "pool1" belongs to subnet 192.172.1.0/24, and "pool2" belongs to subnet 192.168.1.0/24
```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: pool1
  namespace: metallb-system
spec:
  addresses:
    - 192.172.1.10-192.172.1.70

---
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: pool2
  namespace: metallb-system
spec:
  addresses:
    - 192.168.1.128/26
```
In the following YAML, we can configure the VIP in the subnet 192.172.1.0/24 to publish from veth1 (192.172.1.1) and the VIP in the subnet
192.168.1.0/24 published from veth0(192.168.1.1)
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement3
  namespace: metallb-system
spec:
  ipAddressPools:
    - pool1
  interfaces:
    - veth1
---
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement4
  namespace: metallb-system
spec:
  ipAddressPools:
    - pool2
  interfaces:
    - veth0
```

> ### Don't public VIPs to Pods

Reviewer 2: ```suggestion
### Avoid spamming ARP to veth interfaces
```

- Specifies that the VIP is advertised from all interfaces, except the virtual interfaces that connect to the pod in 
  the form of veth pair. For instance, each node in the cluster has 3 physical interfaces: eth0, eth1, eth2, then the YAML like 
  bellow example.
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
name: example-advertisement5
namespace: metallb-system
spec:
  interfaces:
    - eth0
    - eth1
    - eth2
```

> ### Announcing the VIP from specified interfaces of subset of nodes

Reviewer 2: ```suggestion
### Announcing the VIP from specific interfaces only from a subset of nodes
```

```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement6
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
  nodeSelector:
    - matchExpression:
      - key: kubernetes.io/hostname
        operator: In
        values: [hostA, hostB]
  interfaces:
    - ens18
```
### Specify different interfaces for different nodes to announce VIP
- In a heterogeneous cluster, the interfaces of each node are not completely consistent. We need to specify different 
interfaces for different nodes.
For example, listen on eno1 on "worker" nodes, but enp3s5 and vlan6 on "gateway" nodes.
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement7
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
  nodeSelector:
    - role: worker
  interfaces:
    - eno1
---
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement8
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
  nodeSelector:
    - networkRole: gateway
  interfaces:
    - enp3s5
    - vlan6
```

> #### Note: Multiple L2Advertisement are combined by taking the union

Reviewer 2: ```suggestion
#### Note: Multiple L2Advertisements are combined by taking the union
```

apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement10
  namespace: metallb-system
spec:
  ipaddresspools:
    - pool1
  nodeSelector:
    - kubernetes.io/hostname: hostB
  interfaces:
    - ens18
```
The above YAML indicates that MetalLB should advertise the VIPs of all IPAddressPools including pool1 from the interface 
eno1 of all nodes, and also advertise the VIPs in pool1 from ens18 of hostB. In other words, if MetalLB chooses hostB to 
announce the VIP of pool1, the Speaker should announce the VIP from the interfaces ens18 and eno1; if it chooses other nodes, 
the Speaker should announce the VIP only from the interface eno1.

> ### Test Plan

Reviewer: Can you elaborate more about what/how you're going to test?
like this for example: https://github.com/metallb/metallb/blob/main/design/0001-frr.md#test-coverage

Reviewer 2: > Can you elaborate more about what/how you're going to test? like this for example: https://github.com/metallb/metallb/blob/main/design/0001-frr.md#test-coverage

@qingwusunny ping, not sure you missed this comment

Author: Sorry, I have been busy recently. And I will reply later.

Reviewer 2: No rush at all! I wanted just to make sure you did not miss it, thanks again!

# Configuring announcing interfaces

## Summary

> The purpose of this enhancement is to make Speakers only announce VIPs on specified network interfaces.

Reviewer 2: ```suggestion
The purpose of this enhancement is to make the Speakers only announce VIPs on specified network interfaces.
```

Previous discussions about this topic can be found in [this issue](https://github.com/metallb/metallb/issues/277).

## Motivation

A system may have multiple networks based on functions, and thus people want MetalLB only works on specified networks.

> In BGP mode, this situation can be handled by BGP configuration, while in L2 mode,  there is no way to achieve this purpose.

Reviewer 2: nit: multiple spaces

[<img src="./images/example_system.png" width="800"/>](./images/example_system.png)

### Goals

> - The network interfaces the MetalLB uses to broadcast VIP can be specified.

Reviewer 2: I'd specify "for L2 mode"

### Non-Goals
- Change any core functions of MetalLB

## Data model

Based on the current MetalLB CRD design (`8d906f0abef3d7ed5b0435fa4edfbe5e37b38557`), we'd like to add a new field `BindInterfaces` to `L2Advertisement`.

> `BindInterfaces` describes network interface names of the Nodes MetalLB should use. It is a list of interfaces' names, and it is optional. Once users set "bindInterface", then Speakers should only create announcers on these interfaces. Otherwise, Speakers keep the default behavior which creates announcers on every interface.

Reviewer 2: ```suggestion
`BindInterfaces` describes network interface names of the nodes MetalLB should use. It is a list of interfaces names, and it is optional. Once users set "bindInterface", then the Speakers should only create announcers on these interfaces. Otherwise, the Speakers keep the default behavior which creates announcers on every interface.
```

The new definition of `L2Advertisement` is:
```go
type L2Advertisement struct {
  Name                     string `yaml:"name"`
  IPAddressPools           []string `yaml:"ipAddressPools"`
  IPAddressPoolSelector    *metav1.LabelSelector `yaml:"ipAddressPoolSelector"`
  BindInterfaces           []string `yaml:"bindInterfaces"`
}
```

An example definition can be:
```yaml
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: example-advertisement
  namespace: metallb-system
spec:
    ipaddresspools:
      - pool1
      - pool2
    bindInterfaces:
      - eth0
      - eth1

> ```

Reviewer 2: Can you extend the example to see if we are able to support the more complex scenario described in https://github.com/metallb/metallb/issues/277#issuecomment-559599874 ? 
I think we'd be (more or less for free), given the L2Advertisement node selector.

Author: Sorry, what does the second line mean？Can you describe it more detailly?

Reviewer 2: If you look at https://github.com/metallb/metallb/issues/277#issuecomment-559599874, Dave specified the fact that we want to be able to handle both easy scenarios, where all the nodes are equal
```
layer2-config:
  interfaces: ["eno1", "eno2"]
address-pools:
  - name: default
    mode: layer2
    addresses:
      - 10.0.0.10 - 10.0.0.100
 ```

but also more complex scenarios where we want different configs based on the nodes:
```
address-pools:
  - name: default
    protocol: layer2
    addresses:
      - 10.0.0.5 - 10.0.100.10
    interfaces:
      - interfaces: ["eno1", "vlan6"]
        node-selectors:
          - match-labels:
              rack-name: foo
          - match-labels:
              rack-name: bar
      - interfaces: ["eno2"]
        node-selectors:
          - match-labels:
              rack-name: quux
```

Now, this being a property of L2Advertisement, and given the fact that L2Advertisement has a node selector, I _think_ we are covered in that regard. However, I'd document it explicitly (about the use case), and also I'd ensure it works with e2e tests.

Author: Thanks! @fedepaol @gclawes 
I reanalyzed my scenario in the morning.

![veth](https://user-images.githubusercontent.com/26404219/168544685-0b58a8cc-d9f3-4ac3-9d76-711d9422c95b.svg)

The above image describes my initial scenario. The fatal problem is the failure of network connectivity due to asymmetric routing.

For example:
```
 veth0 management network on 192.0.2.0/24
 veth1 data network on 198.51.100.0/24
```
Because MetalLB advertises all IPs from all interfaces, the management client may send a request to veth1, LB Service will reply from veth0 according to the route, which constitutes asymmetric routing so that the host will drop the packets sent by the management client.

My real need is for different subnet IPs to be advertised by the interface specified in node. Subnet IPs can be described by AddressPool. So the design goal is different AddressPools to be advertised by the interface specified in the specified node.
If you agree with that, I will revise my design according to the above goals. @fedepaol @gclawes

## Details

> When the property `bindInterfaces` of L2Advertisement changed, Speaker will close or open ARP and NDP responders according to the "bindInterface".

Reviewer 2: ```suggestion
When the property `bindInterfaces` of L2Advertisement changes, the Speaker will close or open ARP and NDP responders according to the "bindInterface".
```

> When the property `bindInterfaces` of L2Advertisement changed, Speaker will close or open ARP and NDP responders according to the "bindInterface".

Reviewer 2: What happens when we specify a selector that doesn't include any interfaces on a given node?
Do we have any plans?

Author: In this scenario, the Speaker will not open any responders. Maybe we can add a condition type of CRD status for it.

Reviewer 2: Let's be explicit about the fact that the absence of a valid interface won't influence the way we elect the speaker for a given ip.

Author: I has explicit it in the "Non-Goals" part of the second commit
